resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh

```diff
@@ -30,6 +30,10 @@ set -e
 # If there is no passwd entry for the container UID, attempt to create one
 if [ -z "$uidentry" ] ; then
     if [ -w /etc/passwd ] ; then
+        # TODO Should we allow providing an environment variable to set the desired username?
```

Comment:
I think it's ok to remove this TODO. The user name propagation in Hadoop should be fixed elsewhere (in fact I'm touching that code in another PR...).

But more interesting, by having a default UID, this code becomes a no-op, since `/etc/passwd` is not writable anymore.

And that's also probably the cause of the test failure you're seeing.

So maybe you need to call `useradd` or whatever is the appropriate command in this image to create that user when creating the image.

Reply:
Not the case.  `USER` directives take effect at the point at which they occur in the `Dockerfile`, so since the directive is placed after the `ENTRYPOINT` directive the entry point script still runs as `root` allowing adding UID entries and any other privileged setup operations that are needed prior to the containers main process running and having its UID set appropriately

If you manually drop into the resulting container you can see that your UID is the specified non-privileged UID but that an `/etc/passwd` entry was successfully added for you by the entry point script

Reply:
So aren't you basically saying that the entrypoint script is being run a root? Isn't that exactly what we do not want, since that will be running user code as root inside the image?

Reply:
You can run the user code with a securityContext where you can specify runAs: {UID}, but without root you are unable to run useradd commands which would be crucial for said feature. Kubernetes defaults the security context to be root.
Also, is there a security problem with running as root in an isolated container?

Reply:
@rvesse I tried your patch with a modified entrypoint just to print the UID. Here's what it prints:

```
AT START: 185
AT RUN POINT: 185, CMD: 
```

The patch:

```
--- a/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh
+++ b/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh
@@ -19,6 +19,8 @@
 # echo commands to the terminal output
 set -ex
 
+echo "AT START: $(id -u)"
+
 # Check whether there is a passwd entry for the container UID
 myuid=$(id -u)
 mygid=$(id -g)
@@ -91,6 +93,10 @@ if ! [ -z ${HADOOP_CONF_DIR+x} ]; then
   SPARK_CLASSPATH="$HADOOP_CONF_DIR:$SPARK_CLASSPATH";
 fi
 
+echo "AT RUN POINT: $(id -u), CMD: $SPARK_K8S_CMD"
+grep $(id -u) /etc/passwd
+exit 0
+
 case "$SPARK_K8S_CMD" in
   driver)
     CMD=(
```

So I think two things from that experiment:

- what I said is right, that the entrypoint is being run as the defined UID even if it's declared before the UID switch, and that code is a no-op if you defined the UID.
- docker seems to create the UID for you if it doesn't exist. either that or the image we're inheriting from already has that UID.

Reply:
@ifilonenko 
> is there a security problem with running as root in an isolated container?

Other than it always being a horrible idea to run things as root, some deployments may disable containers that run as root from working. And if they do, and you mount some host directory into the container, the container has full privileges over that directory, which can be an issue.

In general, it's better to just avoid doing things as root since it raises more questions than it answers.

Reply:
@vanzin Ok I was wrong here, you are correct that the UID set by the `USER` directive is in effect when the entry point runs.

This section still works because we don't see a UID so the user still belongs to the `root` group which has write access to the `/etc/passwd` file:

```
bash-4.4$ ls -l /etc/ | grep pass
-rw-rw-r--    1 root     root          1281 Nov 14 17:38 passwd
```

Reply:
Actually my previous analysis is not completely correct. The entrypoint is indeed running as user 185. But:

```
bash-4.4$ ls -l /etc/passwd
-rw-rw-r--    1 root     root          1272 Nov 14 17:40 /etc/passwd
```

And:

```
bash-4.4$ id -u
185
bash-4.4$ id -g
0
```

So the group is still `root`, and user 185 can actually do a lot of things that only root should.

Reply:
@vanzin Well I am mostly following advice via @skonto from https://docs.okd.io/latest/creating_images/guidelines.html#use-uid which says:

> For an image to support running as an arbitrary user, directories and files that may be written to by processes in the image should be owned by the root group and be read/writable by that group. Files to be executed should also have group execute permissions.
>
>Adding the following to your Dockerfile sets the directory and file permissions to allow users in the root group to access them in the built image:
> ...
>Because the container user is always a member of the root group, the container user can read and write these files. The root group does not have any special permissions (unlike the root user) so there are no security concerns with this arrangement. In addition, the processes running in the container must not listen on privileged ports (ports below 1024), since they are not running as a privileged user.

Which is also where this code snippet/pattern originates from the folks who originally developed the container images for Spark on K8S

Reply:
> Because the container user is always a member of the root group

I guess that settles it. So what you have is fine. (Comment about the TODO not being needed still applies, though.)